sql/collations.go

```diff
@@ -0,0 +1,857 @@
+// Copyright 2022 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sql
+
+import (
+	"fmt"
+	"strings"
+
+	"gopkg.in/src-d/go-errors.v1"
+
+	"github.com/dolthub/go-mysql-server/internal/regex"
+)
+
+var ErrCollationNotSupported = errors.NewKind("Unknown collation: %v")
+
+// Collation represents the collation of a string.
+type Collation struct {
+	ID           CollationID
+	Name         string
+	CharacterSet CharacterSetID
+	IsDefault    bool
+	IsCompiled   bool
+	SortLength   uint8
+	PadAttribute string
+	compare      collationCompare
+	like         collationLike
+}
+
+// CollationsIterator iterates over every collation available, ordered by their ID (ascending).
+type CollationsIterator struct {
+	idx int
+}
+
+type insensitiveMatcher struct {
+	regex.DisposableMatcher
+}
+
+type collationCompare byte
+type collationLike byte
+
+const (
+	collationCompareInsensitive collationCompare = iota
+	collationCompareSensitive
+)
+
+const (
+	collationLikeInsensitive collationLike = iota
+	collationLikeSensitive
+)
+
+var collationStringToID = map[string]CollationID{}
+
+// CollationID represents the collation's unique identifier. Maybe safely be converted to and from an uint16 for storage.
+type CollationID uint16
+
+const (
+	Collation_armscii8_general_ci         CollationID = 32
```

Comment:
Are the integers meaningful? Why are they ordered as they are? Can we expect this to be stable over time? Some comments would be helpful